properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+"""Classes for dealing with HasProperties instances with unique IDs"""
+import uuid
+
+from six import string_types, text_type
+
+from .. import base, basic, handlers, utils
+from ..base.instance import CLASS_TYPES
+
+
+class HasUID(base.HasProperties):
+    """HasUID is a HasProperties class that includes unique ID
+
+    Adding a UID to HasProperties allows serialization of more complex
+    structures, including recursive self-references. They are serialized
+    to a flat dictionary of UID/HasUID key/value pairs.
+    """
+
+    _REGISTRY = dict()
+    _INSTANCES = dict()
+
+    uid = basic.String(
+        'Unique identifier',
+        default=lambda: text_type(uuid.uuid4()),
+    )
+
+    def __init__(self, **kwargs):
+        super(HasUID, self).__init__(**kwargs)
+        self._INSTANCES[self.uid] = self
+
+    @handlers.validator('uid')
+    def _ensure_unique(self, change):
+        if self.uid == change['value']:
+            pass
+        elif change['value'] in self._INSTANCES:
+            raise utils.ValidationError(
+                message='Uid already used: {}'.format(change['value']),
```

Comment:
"Uid", "uid" or "UID"? Thinking we should be consistent.

Reply:
Yeah, should be consistent.